libyul/ObjectCache.cpp

```diff
@@ -0,0 +1,141 @@
+#include <libyul/optimiser/Suite.h>
+
+#include <liblangutil/DebugInfoSelection.h>
+
+#include <libsolutil/Keccak256.h>
+
+#include <boost/algorithm/string.hpp>
+
+#include <numeric>
+
+using namespace solidity;
+using namespace solidity::langutil;
+using namespace solidity::util;
+using namespace solidity::yul;
+
+
+Dialect const& yul::languageToDialect(Language _language, EVMVersion _version)
+{
+	switch (_language)
+	{
+	case Language::Assembly:
+	case Language::StrictAssembly:
+		return EVMDialect::strictAssemblyForEVMObjects(_version);
+	case Language::Yul:
+		return EVMDialectTyped::instance(_version);
+	}
+	util::unreachable();
+}
+
+void ObjectCache::optimize(Object& _object, Context const& _context, bool _isCreation)
+{
+	yulAssert(_object.code);
+	yulAssert(_object.analysisInfo); // TMP: Do I need this?
+	yulAssert(_object.debugData);
+
+	for (auto& subNode: _object.subObjects)
+		if (auto subObject = dynamic_cast<Object*>(subNode.get()))
+		{
+			bool isCreation = !boost::ends_with(subObject->name, "_deployed");
+			optimize(
+				*subObject,
+				_context,
+				isCreation
+			);
+		}
+
+	Dialect const& dialect = languageToDialect(_context.language, _context.evmVersion);
+	std::unique_ptr<GasMeter> meter;
+	if (EVMDialect const* evmDialect = dynamic_cast<EVMDialect const*>(&dialect))
+		meter = std::make_unique<GasMeter>(*evmDialect, _isCreation, _context.expectedExecutionsPerDeployment);
+
+	h256 cacheKey = calculateCacheKey(*_object.code, *_object.debugData, _context, _isCreation);
+	if (m_cachedObjects.count(cacheKey) != 0)
+	{
+		retrieveObject(cacheKey, _object);
+		return;
+	}
+
+	OptimiserSuite::run(
+		dialect,
+		meter.get(),
+		_object,
+		// Defaults are the minimum necessary to avoid running into "Stack too deep" constantly.
+		_context.optimizeStackAllocation,
+		_context.yulOptimiserSteps,
+		_context.yulOptimiserCleanupSteps,
+		_isCreation ? std::nullopt : std::make_optional(_context.expectedExecutionsPerDeployment),
+		{}
+	);
+
+	storeObject(cacheKey, _object);
+}
+
+void ObjectCache::storeObject(util::h256 _cacheKey, Object& _object)
+{
+	m_cachedObjects[_cacheKey] = CachedObject{
+		std::make_shared<Block>(ASTCopier{}.translate(*_object.code)),
+		_object.analysisInfo,
+		_object.debugData,
+	};
+}
+
+void ObjectCache::retrieveObject(util::h256 _cacheKey, Object& _object) const
+{
+	solAssert(m_cachedObjects.count(_cacheKey) != 0);
+	_object.code = std::make_shared<Block>(ASTCopier{}.translate(*m_cachedObjects.at(_cacheKey).ast));
+	_object.analysisInfo = m_cachedObjects.at(_cacheKey).analysisInfo;
+	_object.debugData = m_cachedObjects.at(_cacheKey).debugData;
+	solAssert(_object.code);
+	solAssert(_object.analysisInfo);
+	solAssert(_object.debugData);
+}
+
+h256 ObjectCache::calculateCacheKey(
+	Block const& _ast,
+	ObjectDebugData const& _debugData,
+	Context const& _context,
+	bool _isCreation
+)
+{
+	AsmPrinter asmPrinter(nullptr, _debugData.sourceNames, DebugInfoSelection::All());
```

Comment:
The failing tests for pure Yul compilation are caused by this. Without providing the original source the debug info does not get printed. This shows up most easily for Yul since the source locations are then pointing at Yul source but I think it should be reproducible at Solidity level too.

I can see a few solutions:
1. Adjust asm printer to make it possible to print locations even without source available. This should be possible as long as we don't request the snippets.
2. Do not use cache for pure Yul (my initial idea, but if it's reproducible at Solidity level then it's a no-go).
3. Store the original source along with the cached object.
4. Switch to calculating cache key directly from AST and account for debug info properly, directly in the form it's stored.

If I can't do 2., then I think 1. is the way to go here. Still relatively quick to do, not a hack and keeps the design reasonable.

Reply:
Turns out that it wasn't caused by not providing sources. It was instead the fact that `Object::debugData` is empty if the input did not have the `@use-src` annotation, and `AsmPrinter` does not print source locations in that case. I solved the problem by disabling caching for such objects.

A better solution would be if we could ensure that `AsmPrinter` always prints all the debug info it has, but when I tried that it resulted with pretty verbose debug annotations being to Yul source that originally did not have them. I need to dig more into this to check why and if we can work around that. But for the purpose of this PR the current solution is sufficient.

Reply:
We discussed it today and the conclusion was that locations present in debug info when `@use-src` is missing are wrong. They should be from optimized source, not from unoptimized. If that was the case they would not be changing depending on where the unoptimized object was originally located in the source.

I actually fixed that exact problem recently but only for `CompilerStack`, not pure Yul. Now I have a more comprehensive fix: #15309. I rebased the PR on top of it and I can confirm that this fixes the problem for objects that do not have `@use-src`.